add PrimaryWorkspace concept in OOP

heejaechang opened the pull request:
**Customer scenario**

Code lens, LUT, FSA will run more efficient in OOP. also FAR or GOTO now can run as fast as they used to be in OOP.

**Bugs this fixes:** 

No bug tracking this issue.

**Workarounds, if any**

Scenario already works just not as efficient as with this change.

**Risk**

quite a bit of changes on how solution is created and maintained in OOP. so there is a bit of risk. it would be better for us to start dogfood sooner.

**Performance impact**

this should improve perf (elapsed time) on all features that run on OOP. since having primary workspace means we now have correct versions, declaration only compilations are cached and etc. especially on features that require versions to work correctly such as FAR, GOTO, Code lens.

this should also improve overall memory usage of OOP since we no longer need to cache all data in memory as we used to. we just need them for initial solution and can get rid of them.

**Is this a regression from a previous update?**

no

**Root cause analysis:**

this is not a regression. this is a work to improve existing code's performance.

**How was the bug found?**

main issue with previous implementation was that, it always re-created solution from scratch. so FAR, GOTO, CodeLens couldn't use cache incrementally. now, those features should be able to. versions in OOP is not same as ones in InProc but, versions should be consistent within OOP itself.

heejaechang commented:
@CyrusNajmabadi can you take a look?

heejaechang commented:
retest this please

heejaechang commented:
all tests seem stuck. retesting.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Workspace/Solution/ProjectState_Checksum.cs:
```diff
@@ -5,6 +5,8 @@
 using Microsoft.CodeAnalysis.Serialization;
 using Microsoft.CodeAnalysis.Internal.Log;
+using System.Runtime.CompilerServices;
+using System;
```
sort

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Workspace/Solution/ProjectState_Checksum.cs:
```diff
@@ -54,5 +62,37 @@ public bool TryGetStateChecksums(out ProjectStateChecksums stateChecksums)
+
+            public static Checksum GetOrCreate(object value, Func<Checksum> checksumCreator)
+            {
+                object saved;
```
why is this not just calling ```return GetOrCreate<Checksum>(value, checksumCreator);```?

CyrusNajmabadi commented on src/Workspaces/Remote/Core/Services/AssetStorage.cs:
```diff
@@ -18,8 +18,8 @@ internal class AssetStorage
-        private const int CleanupInterval = 3; // 3 minutes
-        private const int PurgeAfter = 30; // 30 minutes
+        private const int CleanupInterval = 1; // 1 minutes
+        private const int PurgeAfter = 3; // 3 minutes
```
1. These feel like these should absolutely not be constants.
2. We can make them non-constants and use useful types like TimeSpan.FromMinutes(...)
3. 3-minutes seems *super short*.  If i get up and go use the restroom or get a coffee, will i face a total purge/sync the moment i get back?

heejaechang replied:
now these are just temp storage until solution is updated. primary workspace itself is now the one that hold onto data.

heejaechang replied:
and these are cache if data is expired, it will be re-fetched from VS. and since most, if not all, is in primary workspace, these data will be rarely actually used. such as when solution is first created or temporary solution from things like preview window is needed.

CyrusNajmabadi commented on src/Workspaces/Remote/Core/Services/RemoteWorkspace.cs:
```diff
@@ -0,0 +1,144 @@
+        public override bool CanApplyChange(ApplyChangesKind feature)
+        {
+            // apply change is not allowed
+            return false;
```
might be an issue if we have features that do light-up based on if they think they can change the workspace or not.

heejaechang replied:
yep. we might change it to true once we support applying changes in OOP such as supporting code fix in OOP. but we can create new workspace for those case rather than using primary workspace. not sure how to approach yet. but we can think thorough once we start working to support code fix in OOP.

CyrusNajmabadi commented on src/Workspaces/Remote/Core/Services/RemoteWorkspace.cs:
```diff
@@ -0,0 +1,144 @@
+            return false;
+        }
+
+        public override bool CanOpenDocuments
```
=> notation.

heejaechang replied:
sure

CyrusNajmabadi commented on src/Workspaces/Remote/Core/Services/SolutionService.cs:
```diff
@@ -1,14 +1,13 @@
+<<<<<<< HEAD
+=======
 using System.Collections.Generic;
+>>>>>>> bed3959c19858e55cc0c7fca9601cf0034b61f72
```
oops :)

CyrusNajmabadi commented on src/Workspaces/Remote/Core/Services/TemporaryWorkspace.cs:
```diff
@@ -0,0 +1,38 @@
+
+        public override bool CanApplyChange(ApplyChangesKind feature)
+        {
+            // apply change is not allowed
```
this comment isn't helpful.  It states 'what', rather than 'why'.

CyrusNajmabadi commented on src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_NavigateTo.cs:
```diff
@@ -9,8 +9,7 @@ namespace Microsoft.CodeAnalysis.Remote
     {
-        public async Task<SerializableNavigateToSearchResult[]> SearchDocumentAsync(
-             SerializableDocumentId documentId, string searchPattern, byte[] solutionChecksum)
+        public async Task<SerializableNavigateToSearchResult[]> SearchDocumentAsync(SerializableDocumentId documentId, string searchPattern)
```
wrap long lines.

heejaechang replied:
sure

Pilchie commented:
Is this still relevant?

heejaechang commented:
I am working on it.

heejaechang commented on src/Workspaces/Remote/Core/Services/TemporaryWorkspace.cs:
```diff
@@ -0,0 +1,38 @@
+namespace Microsoft.CodeAnalysis.Remote
+{
+    /// <summary>
+    /// this let us have isolated workspace services between solutions such as option services
```
option is stored in workspace. which mean if one API call changes options, all other API call gets the changed options.

temporary workspace make sure options are isolated between API calls.

I think it is current limitation of our in proc workspace. where if one changes option, it will affect all services while they are running. basically our data API such as intellisense provider, extract method service and etc are not actually stateless (same result with same input) since if somebody changes option in completely unrelated code path, it affects that API.

temporary workspace basically ensure that doesn't happen.

heejaechang commented on src/Workspaces/Remote/Core/Services/RemoteWorkspace.cs:
```diff
@@ -0,0 +1,144 @@
+        /// </summary>
+        public Solution AddSolution(SolutionInfo solutionInfo)
+        {
+            if (solutionInfo == null)
```
I made it to use lock just in case. since like adhocworkspace, this workspace is controlled by us, there is only one who can update the workspace. but just to be safe. added lock.

heejaechang commented:
@mavasani @panopticoncentral @KevinH-MS @srivatsn @CyrusNajmabadi @rynowak can you take a look?

heejaechang commented:
@CyrusNajmabadi changes are almost same as what you code reviewed before except I addressed your feedback and added some unit tests to ensure incremental update.

heejaechang commented:
@MattGertz can I get approval for RC3? this is to make OOP more efficient than before and to allow features such as FAR, GOTO to work properly on OOP.

heejaechang commented:
@KevinH-MS @rynowak @ManishJayaswal this contains breaking change. I removed the magic solution checksum from InvokeAsync arguments

MattGertz commented:
@heejaechang OOP always makes me nervous. :-) Please make private bits available to our vendors and ask them to pound on it for an hour or so with manual scenarios.  Once that's done & shown to be OK, and the CRs are in, then I'll feel OK to approve.  (Tag me again once that's all done.)

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.SolutionChecksumUpdater.cs:
```diff
@@ -60,11 +59,8 @@ private class SolutionChecksumUpdater : GlobalOperationAwareIdleProcessor
-                    await SynchronizeAssets().ConfigureAwait(false);
-                }
+                // update primary solution in remote host
+                await SynchronizePrimaryWorkspace().ConfigureAwait(false);
```
Method should end with Async.

heejaechang replied:
oops.

CyrusNajmabadi commented on src/VisualStudio/Core/Test.Next/Services/AssetStorageTests.cs:
```diff
@@ -47,17 +43,30 @@ public void TestGetAssets()
+            for (var i = 0; i < 10; i++)
             {
-                throw new NotImplementedException();
+                await Task.Delay(10);
```
Delays in tests suck.  This code potentially delays for 100 seconds.  Why can't we use any of our standard mechanisms for listening for async things to happen in tests?  i.e. Waiters and whatnot?

heejaechang replied:
sure.

heejaechang replied:
actually our IAsyncToken thing is not right one to use here. this is not tracking async operations.

sure it can take up to 100ms to find out test is failed. and that is actually better than waiting for event in failed case since that will cause whole jenkin to time out.

usually worry is on false positive (failing when it is actually working correctly) but since it depends on same task for timing (cache cleanup and wait on test), I dont think false positive will happen. and we can address that if it ever actually happen.

CyrusNajmabadi commented on src/VisualStudio/Core/Test.Next/Services/ServiceHubServicesTests.cs:
```diff
@@ -40,30 +45,256 @@ public void TestRemoteHostConnect()
+            {
+                "class CS { }",
+                "class CS2 { }"
+            }, new string[]
```
can use implicit arrays forall of these cases.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Execution/AbstractReferenceSerializationService.cs:
```diff
@@ -30,6 +34,38 @@ protected AbstractReferenceSerializationService(ITemporaryStorageService storage
+
+            var serialized = reader.ReadByte();
+
+            // portable layer doesn't support serialization
```
why?

heejaechang replied:
I don't know. encoding type in portable layer doesnt support serialization.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Execution/AbstractReferenceSerializationService.cs:
```diff
@@ -507,6 +500,58 @@ private unsafe void WriteTo(MetadataReader reader, ObjectWriter writer, Cancella
+            var image = reference as AnalyzerImageReference;
+            if (image != null)
+            {
+                // TODO: think a way to support this or a way to deal with this kind of situation.
```
file bug on this.

heejaechang replied:
sure

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Remote/WellKnownRemoteHostServices.cs:
```diff
@@ -6,7 +6,7 @@ internal class WellKnownRemoteHostServices
     {
         public const string RemoteHostService = "remoteHostService";
         public const string RemoteHostService_Connect = "Connect";
-        public const string RemoteHostService_SynchronizeAsync = "SynchronizeAsync";
```
can we not use nameof() here to reduce hte number of string cases we need to handle?

heejaechang replied:
not fully understanding your suggestion.

CyrusNajmabadi replied:
See the work i did for my OOP features.  Where i define an interface and then use nameof in one place (where we need the string, and implement the interface in the other place to make sure hte name matches).

CyrusNajmabadi replied:
RIght now, someone could rename "SynchronizePrimaryWorkspaceAsync" and not know to update this location.  The nameof approach addresses that.

heejaechang replied:
but, that can happen in other PR.

CyrusNajmabadi commented on src/Workspaces/Remote/Core/Services/RemoteWorkspace.cs:
```diff
@@ -0,0 +1,160 @@
+        // REVIEW: I am using semaphoreSlim since workspace is using it, but not sure why it uses
+        //         semaphore rather than just object since workspace is not using anything specific
+        //         to semaphore
+        private readonly SemaphoreSlim _serializationLock = new SemaphoreSlim(initialCount: 1);
```
Doesn't the workspace use async blocking on the SemaphoreSlim?  if not, we can change them to be normal objects.

heejaechang replied:
they don't use async dispoablewait API.

ManishJayaswal commented:
@KevinH-MS @heejaechang for vendors to play with this, we will need a private Roslyn build and a corresponding LUT build that they could patch their machines with, right?

KevinH-MS commented:
@ManishJayaswal yes, I can probably produce a LUT build later tonight.

CyrusNajmabadi commented:
> changes are almost same as what you code reviewed before except I addressed your feedback

With 2k lines changed... i'm not going to be able to review all of this thoroughly in one go. I have to do it in passes, focusing on different areas each time.

KevinH-MS commented:
@heejaechang I tried compiling against this, and I didn't see any breaks (except some recent changes in the ```IOptionProvider```).

We are currently calling this overload:  ```public abstract Task<T> InvokeAsync<T>(string targetName, params object[] arguments)```

Should we need to change anything?

heejaechang commented:
@KevinH-MS caller side shouldn't be changed. but your service side need to be changed.
ex) void MyService(mytype argument, byte[] solutionchecksum) -> void MyService(mytype argument)

otherwise, InvokeAsync will not find the end point.

heejaechang commented:
@KevinH-MS @ManishJayaswal  I have signed bit I created in cpvsbuild from this PR under roslyn-signed\head.

heejaechang commented on src/Workspaces/Core/Portable/Execution/AbstractReferenceSerializationService.cs:
```diff
@@ -30,6 +34,38 @@ protected AbstractReferenceSerializationService(ITemporaryStorageService storage
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            writer.WriteByte(NoEncodingSerialization);
```
encoding serialization only supported in desktop.

KevinH-MS commented:
@heejaechang can you send me a link or the build number?

heejaechang commented on src/Workspaces/Core/Desktop/Execution/DesktopReferenceSerializationServiceFactory.cs:
```diff
@@ -22,10 +29,62 @@ public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
+                writer.WriteByte(EncodingSerialization);
+
+                byte[] value;
+                if (!s_encodingCache.TryGetValue(encoding, out value))
```
because I believe 99% time TryGetValue will succeed

CyrusNajmabadi replied:
I'm not understanding.  Why would the func approach be bad there?  It's not like the func approach will allocate.  The lambda will be cached, and we'll only need to call it if the lookup fails.

heejaechang replied:
but why is this bad? already written this way and I dont see reason to change to lambda one?

CyrusNajmabadi replied:
imo, because it's much clearer.  There is a method which expresses exactly the concept you're trying to express.  Namely: give me the value if you have it.  Otherwise, compute this value and put it in the map and give me that.

heejaechang commented:
ping? @panopticoncentral ?

heejaechang commented:
retest windows_build_correctness_prtest please

commit status update failure

heejaechang commented:
close and re open to re-run tests

heejaechang commented:
ping?

heejaechang commented:
ping? can I get some code review? looks like I need 2 sign off

MattGertz commented:
Still waiting on CRs; ping me when you get them.

heejaechang commented:
I got sign off from private bit testing.

@panopticoncentral @mavasani @srivatsn @KevinH-MS @CyrusNajmabadi can i get code review, I need 2 sign off

panopticoncentral commented:
:shipit:

heejaechang commented:
one more sign off?

heejaechang commented:
@mavasani can you take a look?

mavasani commented on src/EditorFeatures/TestUtilities/Remote/InProcRemostHostClient.cs:
```diff
@@ -21,9 +21,9 @@ internal class InProcRemoteHostClient : RemoteHostClient
         private readonly JsonRpc _rpc;
 
-        public static async Task<RemoteHostClient> CreateAsync(Workspace workspace, CancellationToken cancellationToken)
+        public static async Task<RemoteHostClient> CreateAsync(Workspace workspace, bool forTesting, CancellationToken cancellationToken)
```
`forTesting` seems confusing here as this code is already in the test layer.

heejaechang replied:
sure I will change it to runCleanup

mavasani commented on src/VisualStudio/RemoteHostClientMock/Remote/RemoteHostClientFactory.cs:
```diff
@@ -21,7 +21,7 @@ internal class RemoteHostClientFactory : IRemoteHostClientFactory
                 if (workspace.Options.GetOption(RemoteHostClientFactoryOptions.RemoteHost_InProc))
                 {
-                    return await InProcRemoteHostClient.CreateAsync(workspace, cancellationToken).ConfigureAwait(false);
+                    return await InProcRemoteHostClient.CreateAsync(workspace, forTesting: false, cancellationToken: cancellationToken).ConfigureAwait(false);
```
This seems like the only place where we are using `forTesting = false`. Couple of suggestions:

1. Can we move the RemoteHostClientMock under some test folder? Currently the folder structure `src/VisualStudio/RemoteHostClientMock/` makes it seem like it is shipping product code.
2. Can you add comments on why we are setting this flag to false here?

mavasani commented on src/VisualStudio/Core/Test.Next/Mocks/TestAnalyzerAssemblyLoader.cs:
```diff
@@ -0,0 +1,19 @@
+
+namespace Roslyn.VisualStudio.Next.UnitTests.Mocks
+{
+    public class TestAnalyzerAssemblyLoader : IAnalyzerAssemblyLoader
```
Can you instead use [this](http://source.roslyn.io/#Roslyn.Compilers.UnitTests/AnalyzerFileReferenceTests.cs,282373cf1db99c90,references) test-only loader with identical implementation?

mavasani commented on src/Workspaces/Remote/Core/Services/AssetService.cs:
```diff
@@ -107,7 +114,7 @@ private bool EnsureCacheEntryIfExists(Checksum checksum)
         }
 
-        private async Task SynchronizeAssetsAsync(ISet<Checksum> checksums, CancellationToken cancellationToken)
+        public async Task SynchronizeAssetsAsync(ISet<Checksum> checksums, CancellationToken cancellationToken)
```
Should this be call AddAssetsAsync? The name makes it confusing with your newly added method above which actually calls synchronize assets on ChecksumSynchronizer.

mavasani commented on src/Workspaces/Remote/Core/Services/SolutionService.cs:
```diff
@@ -36,21 +31,31 @@ public SolutionService(AssetService assetService)
+                return currentSolution;
+            }
+
+            var lastSolution = s_lastSolution;
```
Can this be invoked from multiple threads?

heejaechang commented on src/VisualStudio/Core/Next/Remote/JsonRpcSession.cs:
```diff
@@ -65,43 +65,29 @@ internal class JsonRpcSession : RemoteHostClient.Session
-                await _snapshotClient.InvokeAsync(WellKnownServiceHubServices.ServiceHubServiceBase_Initialize, _currentSessionId, PinnedScope.SolutionChecksum.ToArray()).ConfigureAwait(false);
-                await _serviceClient.InvokeAsync(WellKnownServiceHubServices.ServiceHubServiceBase_Initialize, _currentSessionId, PinnedScope.SolutionChecksum.ToArray()).ConfigureAwait(false);
-            }
-            catch (ObjectDisposedException)
```
these try/catch is not needed. looks like haven't been cleaned up when try/catch is moved into InvokeAsync

heejaechang commented:
@MattGertz can I get approval for RC3? I got 2 sign off. thank you

heejaechang commented:
@srivatsn it might be you who I need to get sign off? I got 2 sign off and vender did private testing and I did some dogfooding working with this bit doing regular daily work. and all seems working fine.
